packages/bitcore-node/src/services/p2p/index.ts

```diff
@@ -0,0 +1,268 @@
+import { ITransactionModel } from '../../models/transaction';
+import { SupportedChain, SupportedChainSet } from '../../types/SupportedChain';
+import { BtcP2pService } from './bitcoin';
+import { Bitcoin } from '../../types/namespaces/Bitcoin';
+import { setImmediate } from 'timers';
+import { EventEmitter } from 'events';
+
+
+export interface P2pService<Block, Transaction> {
+  // a stream of incoming blocks
+  blocks(): Observable<CompleteBlock<Block, Transaction>>;
+
+  // a stream of incoming transactions
+  transactions(): Observable<Transaction>;
+
+  // connect to the peers and begin emitting data
+  start(): Promise<void>;
+
+  // sync data from previous history according to these hashes
+  // returns the hash of the last block synced
+  sync(locatorHashes: string[]): Promise<string | undefined>;
+
+  // get the max height of every peer in the pool
+  height(): number;
+
+  // get information about the chain this forked from (if applicable)
+  parent(): ChainNetwork & { height: number } | undefined;
+
+  // disconnects from peer and stops all pending tasks,
+  // afterwards `start()` can be called again.
+  stop(): Promise<void>;
+
+  // when `true` only emit blocks that result from the syncing process
+  syncing: boolean;
+}
+
+
+export type StandardP2p = P2pService<Bitcoin.Block, Bitcoin.Transaction>;
+
+
+export type CompleteBlock<B, T> = {
+  block: B,
+  transactions: T[],
+};
+
+
+export enum P2pEvents {
+  SYNC_COMPLETE = 'SYNC_COMPLETE',
+}
+
+
+export class P2pRunner {
+  private service: StandardP2p;
+  private chain: string;
+  private network: string;
+  private blocks: IBlockModel;
+  private transactions: ITransactionModel;
+
+  public events: EventEmitter;
+
+  constructor(
+    chain: string,
+    network: string,
+    blocks: IBlockModel,
+    transactions: ITransactionModel,
+    service: StandardP2p
+  ) {
+    this.service = service;
+    this.chain = chain;
+    this.network = network;
+    this.blocks = blocks;
+    this.transactions = transactions;
+    this.events = new EventEmitter();
+  }
+
+  async start(): Promise<{
+    blocks: Observable<CompleteBlock<Bitcoin.Block, Bitcoin.Transaction>>,
+    transactions: Observable<Bitcoin.Transaction>,
+  }> {
+    logger.debug(`Started worker for chain ${this.chain}`);
+    const parent = this.service.parent();
+    const syncer = await this.sync();
+
+    const blocks = this.service.blocks().pipe(concatMap(async pair => {
+      await this.blocks.addBlock({
+        chain: this.chain,
+        network: this.network,
+        forkHeight: parent? parent.height : 0,
+        parentChain: parent? parent.chain : this.chain,
+        block: pair.block
+      });
+      if (this.service.syncing) {
+        syncer.add(pair.block.hash);
+      }
+      else {
+        logger.info(`Added block ${pair.block.hash}`, {
+          chain: this.chain,
+          network: this.network,
+        });
+      }
+      return pair;
+    })).pipe(share());
```

Comment:
not sure if I get need for share() here. Is this to make the Observable fire off without a subscriber?

Reply:
I've used publish and connect before, is this related? 
 https://www.learnrxjs.io/operators/multicasting/publish.html